python/amici/ode_export.py

```diff
@@ -0,0 +1,2266 @@
+    """
+    Checks if the values for a symbolic variable is passed in the signature
+    of a function
+
+    Arguments:
+    ----------
+    name: str
+        name of the function
+
+    varname: str
+        name of the symbolic variable
+
+    Returns:
+    ----------
+
+    Raises:
+    ----------
+
+    """
+    varname = varname.replace('sparse', '')
+    return name in functions \
+           and re.search(
+                    f'const (realtype|double) \*{varname}[0]*[,)]+',
+                    functions[name]['signature']
+                )
+
+
+class ODEExporter:
+    """
+    The ODEExporter class generates AMICI C++ files for ODE model as
+    defined in symbolic expressions.
+
+    Attributes:
+    ----------
+    codeprinter: symengine.printing.CCodePrinter
+        codeprinter that allows export of symbolic variables as C++ code
+
+    functions: dict
+        carries C++ function signatures and other specifications
+
+    modelName: str
+        name of the model that will be used for compilation
+
+    modelPath: str
+        path to the generated model specific files
+
+    modelSwigPath: str
+        path to the generated swig files
+
+    allow_reinit_fixpar_initcond: bool
+        indicates whether reinitialization of initial states depending on
+        fixedParmeters is allowed for this model
+
+    """
+
+    def __init__(
+            self,
+            ode_model,
+            outdir=None,
+            verbose=False,
+            assume_pow_positivity=False,
+            compiler=None
+    ):
+        """
+        Generate AMICI C++ files for the ODE provided to the constructor.
+
+        Arguments:
+        ----------
+        ode_model: ODEModel
+            ODE definition
+
+        output_dir: str
+            see sbml_import.setPaths()
+
+        verbose: bool
+            more verbose output if True
+
+        assume_pow_positivity: bool
+            if set to true, a special pow function is used to avoid
+            problems with state variables that may become negative due
+            to numerical errors
+
+        compiler: str
+            distutils/setuptools compiler selection to build the python
+            extension
+
+        Returns:
+        ----------
+
+        Raises:
+        ----------
+
+        """
+
+        self.outdir = outdir
+        self.verbose = verbose
+        self.assume_pow_positivity = assume_pow_positivity
+        self.compiler = compiler
+
+        self.codeprinter = CCodePrinter()
+
+        self.modelName = 'model'
+        output_dir = os.path.join(os.getcwd(),
+                                  f'amici-{self.modelName}')
+        self.modelPath = os.path.abspath(output_dir)
+        self.modelSwigPath = os.path.join(self.modelPath, 'swig')
+
+        # Signatures and properties of generated model functions (see
+        # include/amici/model.h for details)
+
+        self.model = ode_model
+
+        self.functions = functions
```

Comment:
We should use a deepcopy here, otherwise there might be trouble when working with multiple ODE exporters. Not sure how often that will happen, but better safe than sorry.